### ISSUE TITLE
`scipy.ncx2.sf` should be monotone decreasing

### ISSUE THREAD

**Issue by Phillip-M-Feldman** [scipy/scipy#8665]
I'm getting unexpected behavior from `scipy.ncx2.sf`.  The survival function should eventually go to zero for large enough inputs, but instead I'm seeing a kind of "foot", i.e., it is bottoming out at a positive value.  The following interactive I/O demonstrates the issue.

### Reproducing code **example:**

(Pdb) v_T_norm[::50]
array([  4.99515382e+00,   1.07617327e+01,   2.31854502e+01,
         4.99515382e+01,   1.07617327e+02,   2.31854502e+02,
         4.99515382e+02,   1.07617327e+03,   2.31854502e+03,
         4.99515382e+03,   1.07617327e+04,   2.31854502e+04,
         4.99515382e+04])
(Pdb) nu
20.0
(Pdb) lam
499.51538166556196
(Pdb) ncx2.sf(v_T_norm[::50], df=nu, nc=lam)
array([  1.00000000e+00,   1.00000000e+00,   1.00000000e+00,
         1.00000000e+00,   1.00000000e+00,   1.00000000e+00,
         6.64666099e-01,   4.16325934e-05,   4.16325934e-05,
         4.16325934e-05,   4.16325934e-05,   4.16325934e-05,
         4.16325934e-05])

**Comment by WarrenWeckesser** [comment]
After a quick look, I can see that the problem is in `scipy.special.chndtr`.  The implementation of `nxc2` doesn't define a survival function, so it inherits the default implementation, which is simply 1 minus the CDF.  The CDF for `ncx2` ultimately calls `scipy.special.chndtr`, and it exhibits the "plateau" behavior:

```
In [32]: from scipy.special import chndtr

In [33]: nu = 20.0

In [34]: lam = 499.5

In [35]: chndtr([500, 750, 1000, 2000, 4000], nu, lam)
Out[35]: array([ 0.33945802,  0.99995706,  0.99995841,  0.99995841,  0.99995841])
```

The next step is to figure out what is wrong with `chndtr`.

**Comment by WarrenWeckesser** [comment]
The problem is in the implementation of `cumchn` in the CDFLIB Fortran code.  I'll have a pull request with a fix once I get some known accurate values to use in unit tests.

**Comment by WarrenWeckesser** [comment]
The function `cumchn` has the same problem as discussed in https://github.com/scipy/scipy/issues/8398.  A series is used to compute the CDF, and the code uses a relative precision of just 1e-5 to decide when a term is small enought that the series can be terminated.  We can improve the result by tightening up that tolerance.  However, even after removing the tolerance entirely and only stopping the series sum when adding the next term doesn't change the current sum, I still get a "plateau" (just not as bad as the existing plateau).

For what it's worth, the Boost implementation of the noncentral chi-square distribution appears to be more sophisticated and doesn't have this problem.  Has there been any thought to replacing parts of CDFLIB with Boost, where possible?  I don't like spending time hacking CDFLIB when I know there is a better open-source (and license-compatible) implementation elsewhere.

**Comment by WarrenWeckesser** [comment]
I just created a pull request that improves the calculation, but it is not a complete fix: https://github.com/scipy/scipy/pull/8673

**Comment by WarrenWeckesser** [comment]
A [question on stackoverflow](https://stackoverflow.com/questions/49769952/accuracy-of-non-central-chi-squared-distribution-implementation-in-scipy) also notes the problem with the accuracy of `ncx2.cdf`.  With the change made in https://github.com/scipy/scipy/pull/8673, that example gives

```
In [4]: ncx2.cdf([1200, 1500, 2000, 5000], 2, 1000).tolist()
Out[4]: 
[0.9986639334269182,
 0.9999999999993727,
 1.0000000000000306,
 1.0000000000000306]
```
The values are more accurate, but (as noted in https://github.com/scipy/scipy/pull/8673), the values still "plateau" at a value that is not 1, and in some cases the plateau value is greater than 1.

**Comment by Phillip-M-Feldman** [comment]
The newer issue should be deleted.

**Comment by MohammadKarimzadeh** [comment]
Just wanted to add that this seems to be not only an accuracy problem but it has significant issues.
```
from scipy.stats import ncx2
import numpy as np
import matplotlib.pyplot as plt
from rpy2.robjects.packages import importr  # for comparison
stats = importr(str('stats'))
x = np.arange(5000, 15000, 1)
scdf = [ncx2.cdf(i,30,i) for i in x]
rcdf = [stats.pchisq(i,30,i)[0] for i in x]
f, ax = plt.subplots(figsize = (14, 10))
ax.plot(x, scdf, c='b', label='from scipy')
ax.plot(x, rcdf, c='r', label='from rpy2')
ax.legend(prop={'size': 30})
```
results:
![scipy_vs_rpy2](https://user-images.githubusercontent.com/47407122/52388184-37ea4000-2a5b-11e9-8a27-1a854468580f.png)

**Comment by Phillip-M-Feldman** [comment]
A change that was implemented last year moved the point at which the bad behavior begins out to the right, but did not eliminate the problem.

**Comment by caos21** [comment]
The fix above proposed for #11474 , seems to solve this issue.

 
![image](https://user-images.githubusercontent.com/12665566/73989775-fb268280-4914-11ea-9f0c-0a861e32271c.png)

**Comment by Phillip-M-Feldman** [comment]
This looks perfect!
